File: backend/dataset/coinChange/coinChange_13.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        ## RC ##
        ## APPROACH : DP ##
        
		## TIME COMPLEXITY : O(N^2) ##
		## SPACE COMPLEXITY : O(N) ##

        dp = [0] + [amount + 1] * amount                                # fill invalid amount i.e amount + 1 for all amounts ranging from 0 to amount.
        for value in range(amount+1):                                   # for each value till amount
            for coin in (coins):                                        # for each coin
                if(coin <= value):                                      # check if this coin can be used make value i.
                    dp[value] = min(dp[value], 1 + dp[value - coin])    # 1 => current coin included, check the min coins required to make remaining value i.e  value - coin                    
        return -1 if(dp[amount] > amount) else dp[amount]

File: backend/CDataset/coinChange/coinChange_13.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        // unbounded Knspascak Problem
        
        int n = coins.size();
        int m = amount;
        
        int t[n+1][m+1];
        
        for(int i=0;i<n+1;i++)
        {
            for(int j=0;j<m+1;j++)
            {
                if(i==0 && j==0)
                {
                    t[i][j]=0;
                }
                else if(i==0)
                {
                    t[i][j]=INT_MAX-100;
                }
                else if(j==0)
                {
                    t[i][j]=0;
                }
                else if(coins[i-1]<=j)
                {
                    t[i][j]=min((1+t[i][j-coins[i-1]]),t[i-1][j]);
                }
                else
                {
                    t[i][j]=t[i-1][j];
                }
            }
            
        }
        return (t[n][m]==INT_MAX-100)?-1:t[n][m];
    }
};
